expr/func_summarize.go

```diff
@@ -0,0 +1,100 @@
+package expr
+
+import (
+	"fmt"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/batch"
+	"github.com/grafana/metrictank/consolidation"
+	"github.com/grafana/metrictank/util"
+	"github.com/raintank/dur"
+	"gopkg.in/raintank/schema.v1"
+)
+
+type FuncSummarize struct {
+	in             GraphiteFunc
+	intervalString string
+	fn             string
+	alignToFrom    bool
+}
+
+func NewSummarize() GraphiteFunc {
+	return &FuncSummarize{fn: "sum", alignToFrom: false}
+}
+
+func (s *FuncSummarize) Signature() ([]Arg, []Arg) {
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgString{val: &s.intervalString, validator: []Validator{IsIntervalString}},
+		ArgString{key: "func", opt: true, val: &s.fn, validator: []Validator{IsConsolFunc}},
+		ArgBool{key: "alignToFrom", opt: true, val: &s.alignToFrom},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncSummarize) Context(context Context) Context {
+	context.consol = 0
+	return context
+}
+
+func (s *FuncSummarize) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	interval, _ := dur.ParseDuration(s.intervalString)
+	aggFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(s.fn))
+
+	var alignToFromTarget string
+	if s.alignToFrom {
+		alignToFromTarget = ", true"
+	}
+	newName := func(oldName string) string {
+		return fmt.Sprintf("summarize(%s, \"%s\", \"%s\"%s)", oldName, s.intervalString, s.fn, alignToFromTarget)
+	}
+
+	var outputs []models.Series
+	for _, serie := range series {
+		var newStart, newEnd uint32 = serie.QueryFrom, serie.QueryTo
+		if len(serie.Datapoints) > 0 {
+			newStart = serie.Datapoints[0].Ts
+			newEnd = serie.Datapoints[len(serie.Datapoints)-1].Ts + serie.Interval
+		}
+		if !s.alignToFrom {
+			newStart = newStart - (newStart % interval)
+			newEnd = newEnd - (newEnd % interval) + interval
+		}
+
+		out := summarizeValues(serie, aggFunc, interval, newStart, newEnd)
+
+		output := models.Series{
+			Target:     newName(serie.Target),
+			QueryPatt:  newName(serie.QueryPatt),
+			Tags:       serie.Tags,
+			Datapoints: out,
+			Interval:   interval,
+		}
+		outputs = append(outputs, output)
+		cache[Req{}] = append(cache[Req{}], output)
+	}
+	return outputs, nil
+}
+
+func summarizeValues(serie models.Series, aggFunc batch.AggFunc, interval, start, end uint32) []schema.Point {
+	out := pointSlicePool.Get().([]schema.Point)
+
+	numPoints := int(util.Min(uint32(len(serie.Datapoints)), (start-end)/interval))
```

Comment:
i'm confused here. the only reason to get the util.Min of these two, is just in case the input series had a low number of points wrt to the requested interval right?
e.g.
1 day worth of 2-hourly data
requesting summarize of 1h.
so this becomes `numPoints = min(12,24) = 12`
then why in the loop below do we increment by interval (1h), but only 12 times? that would cover only a 12h timerange?

Reply:
If (serie.Datapoints[i].Ts < ts+interval) is violated, it doesn't increment i. So it would append NaN's for the inbetween. I'll push a test case that shows this. 

There is a bug though, right now it fails with any agg other than sum because of the check.

Reply:
Ahh handling that case is easy. I do wonder why we check for max but not sum. @shanson7 ?